### PR TITLE
#168131413 An Email should be sent to a user upon successful registration. User should be able to receive a verification link via email upon successful signup

### PR DESCRIPTION
**Pull request by Bahatiroben**
#### What does this PR do?
- it send an email with verification link to the user
#### Description of Task to be completed?
- it sends a verification link to the user upon signup which is valid for one hour
- it sets the user as verified after clicking the link hence able to login
- it prevents anyone from signing in before verifying the email

#### How should this be manually tested?
- clone the repo
- checkout to the branch ft-email-verification-link-168131413
- run npm install to install the dependencies
- run npm test to run the tests
- run npm start:dev to start the development server
- signup with this payload to get a user who is currently not verified.
```
{	"firstName": "Robben",
	"lastName": "Bahati",
	"userEmail": "bahati.robben@andela.com",
	"userPassword": "Root1234"
}
```
- you should get something similar to this:
```
{
    "status": 201,
    "message": "Account has been created successfully",
    "data": {
        "firstName": "Robben",
        "lastName": "Bahati",
        "userEmail": "bahati.robben@andela.com",
        "userRoles": null,
        "userToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImJhaGF0aS5yb2JiZW5AYW5kZWxhLmNvbSIsImZpcnN0TmFtZSI6IlJvYmJlbiIsImxhc3ROYW1lIjoiQmFoYXRpIiwiaWF0IjoxNTY4MjI4ODI3fQ.qpqwCvtH51lxR2-v2pfJ-CsrztucNE3uzghuxAF5K20",
        "verification": {
            "message": "Verification link sent",
            "link": "http://localhost:PORT/api/v1/auth/verify/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImJhaGF0aS5yb2JiZW5AYW5kZWxhLmNvbSIsImZpcnN0TmFtZSI6IlJvYmJlbiIsImxhc3ROYW1lIjoiQmFoYXRpIiwiaWF0IjoxNTY4MjI4ODI3fQ.qpqwCvtH51lxR2-v2pfJ-CsrztucNE3uzghuxAF5K20"
        }
    }
}
```

- click the link sent into your email. You should get something like this
```
{
"status": 201,
"message": "Email verified succesfully",
"userEmail": "bahati.robben@andela.com"
}
```
- if you click it again you should get something like this:
```
{
"status": 409,
"error": "Email already verified"
}
```
- if the link is broken, or you delayed to verify(1 hr) you should get something like this:
```
{
"status": 401,
"message": "The link is invalid or has expired"
}
```
- Incase you didn't receive the link, just hit this enadpoint 'api/v1/auth/createLink'
```{
"email": "bahati.robben@andela.com"
}
```
- you should get something similar to this:
```
{
    "status": 200,
    "message": "email sent with verification link",
    "data": {
        "userEmail": "bahati.robben@andela.com",
        "link": "http://localhost:4500/api/v1/auth/verify/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NTEsImVtYWlsIjoiYmFoYXRpLnJvYmJlbkBhbmRlbGEuY29tIiwiZmlyc3ROYW1lIjoiUm9iYmVuIiwibGFzdE5hbWUiOiJCYWhhdGkiLCJpYXQiOjE1NjgyMzE2MDN9.LPFDlTbi977Uz1WYQImUSuNzCLLh4cdnYr355xnduX4"
    }
}
```

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#168131413](https://www.pivotaltracker.com/story/show/168131413)
#### Screenshots (if appropriate)
image
image
image
image
image

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/userController.js [hunk 1]
```diff
@@ -40,7 +41,75 @@ class Users {
       userRoles: data.userRoles,
       userToken: data.token
     };
-    return Response.customResponse(res, 201, 'Account has been created successfully', dataResponse);
+    const link = `${req.protocol}://${process.env.baseUrl}/api/v1/auth/verify/?token=${token}`;
+    let verification;
```
Comment (Dkabiswa) [inline src/controllers/userController.js]: Here you sending a verification email

File: src/controllers/userController.js [hunk 2]
```diff
@@ -40,7 +41,75 @@ class Users {
...
+    const { userEmail } = req.body;
+    const user = await userService.findUserByEmail(userEmail);
+    if (user === null) {
+      return res.status(404).send({
+        status: res.statusCode,
+        message: 'this email is not registered'
+      });
+    }
+
+    const token = await SessionManager.generateToken(user);
```
Comment (Dkabiswa) [inline src/controllers/userController.js:116]: your sending a verification  email here too, whatif you create a method in this class which goes through the whole process of setting URL to sending an email and call it using `this.method` in these required moments

File: src/routes/api/users.js [hunk 3]
```diff
@@ -6,5 +6,7 @@ const router = express.Router();
 
 router.post('/signup', Users.createUser);
 router.post('/signin', userValidator.validateSignIn, Users.logIn);
+router.get('/createLink', userValidator.validateSendLink, Users.sendLink);
+router.use('/verify/', userValidator.validateVerifyLink, Users.verify);
```
Comment (Dkabiswa) [inline src/routes/api/users.js]: i think router.patch for user verification will be more realistic since you updating a database column of verify

**Review by Dkabiswa — CHANGES_REQUESTED** [review]

File: src/controllers/userController.js (hunk 2 above)
Comment (Bahatiroben) [reply src/controllers/userController.js:116]: I thought about it, but it is being hard for me, since it can mess up signup feature

Comment (Dkabiswa) [reply src/controllers/userController.js:116]: its ok we shall refactor code after merging

**Review by Dkabiswa — APPROVED** [review]

File: src/utils/email.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ */
+class Emails {
+  /**
+   * Generates reset password email template
+   * @param {string} url reset password link
+   * @param {Object} user user details
+   * @returns {String} reset password email template
+   */
+  static verificationLinkTemplate(url, user) {
+    const from = 'marveldev53@gmail.com';
```
Comment (Bahatiroben) [inline src/utils/email.js]: insert this in an environment variable

File: src/controllers/userController.js [hunk 5]
```diff
@@ -40,9 +41,74 @@ class Users {
...
+   * @param {object} req  details.
+   * @param {object} res  details.
+   * @returns {object}.
+   */
+  async verify(req, res) {
+    const { token } = req.query;
+    try {
+      const { email } = SessionManager.verify(token);
+      const user = await userService.findUserByEmail(email);
+      if (user.accountVerified === true) {
```
Comment (Bahatiroben) [inline src/controllers/userController.js]: user.isverified

**Review by j0flintking02 — APPROVED** [review]
LGTM

**Pull request merged**